### DISCUSSION TITLE
OS for RVCop64

### DISCUSSION THREAD

**Discussion by pottendo** [zeldin/OrangeCart#3] [Ideas]
hi,
I'm thinking of getting a typical embedded OS running on the RVCop64.

This could be a good starting point:
  https://docs.zephyrproject.org/latest/boards/riscv/litex_vexriscv/doc/index.html#

I managed to build Zephyr already for a RiscV arch and got it running with emulation (QEmu) on my Linux host. Next step would be a 'board support package' for our orange cart - need to continue my research to better understand where to start; there are some guidelines for Zephyr I need to study.

Any feedback or inputs/tips/suggestions are welcome,
pottendo

**Comment by pottendo** [comment]
...signs of life:
[...]
--============== Boot ==================--
Booting from serial...
Press Q or ESC to abort boot completely.
sL5DdSMmkekro
[LITEX-TERM] Received firmware download request from the device.
[LITEX-TERM] Uploading build/zephyr/zephyr.bin to 0x40000000 (33756 bytes)...
[LITEX-TERM] Upload calibration... (inter-frame: 10.00us, length: 64)
[LITEX-TERM] Upload complete (58.6KB/s).
[LITEX-TERM] Booting the device.
[LITEX-TERM] Done.
Executing booted program at 0x40000000

--============= Liftoff! ===============--
*** Booting Zephyr OS build zephyr-v3.2.0-843-gd832b04e9627  ***
Hello World! orangecart

[...]

I have no clue, what I'm doing - but at least the uart console seems to work after fiddling with the devicetree spec...

**Reply by pottendo** [reply]
Some progress:
- Setting the timer0 address properly, even time passes on RVCop64; although a bit too slow (~20%), probably the some frequency setting in the device-tree needs adjustment.
- Multithreading just works: https://photos.app.goo.gl/MUS9NuWWoAftr1K3A
- Access to C64 memory works (probably with the same effects as reported here: [https://github.com/zeldin/OrangeCart/discussions/2])

Todo: 
- map further I/Os (hard to test for me)
- Fix time-passing to match realtime
- Check debugging support(gdb-stub)
- Cleanup and ask for merging into Zephyr... ;-)
(somehow this was too easy!)

**Reply by zeldin** [reply]
Cool.  Did you generate a `csr.json` and convert it to an overlay DTS using `litex/litex/tools/litex_json2dts_zephyr.py` like they suggested on that page you linked?  I would have assumed that would pick the value of `config_clock_frequency` from the JSON to get the correct timer speed, but 20% slow sounds more like the DTS specifies the clock speed to 100 MHz (assuming you are still running at 80 MHz)...

**Reply by zeldin** [reply]
This seems to be the DTS entry that needs overriding for correct speed: https://github.com/zephyrproject-rtos/zephyr/blob/fab85afc911db253bf283ce2868f6600eb379f01/dts/riscv/riscv32-litex-vexriscv.dtsi#L22
As expected it hardcodes 100 MHz.  So you'll need an overlay saying something like
```
/cpus/cpu@0 { clock-frequency = <80000000>; };
```
(or whatever your clock frequency is).

**Reply by zeldin** [reply]
Ah, it seems that Zephyr does not take the base frequency from the DTS file at all.  Anyway, I've now pushed a commit that automatically generates `overlay.dts` and `overlay.conf` in the `hw/build` directory, with the correct settings for your build (including the correct clock frequency).  This allows you to build Zephyr with the `litex_vexriscv` BSP like described on the bottom of this page: https://github.com/enjoy-digital/litex/wiki/Run-Zephyr-On-Your-SoC

**Reply by pottendo** [reply]
hi,
checked a bit around the auto-generation of the RVCop64 specific DTS.
So the file generated by litex_json2dts_zephyr seems pretty good, but not exactly matching - at least isn't compatible with the working console driver in Zephyr: zephyr/drivers/serial/uart_liteuart.c
Some registers are a bit different: from the 'csr.json':
`
        "uart_tuning_word": {
            "addr": 4026535960,
            "size": 1,
            "type": "rw"
        },
        "uart_configured": {
            "addr": 4026535964,
            "size": 1,
            "type": "rw"
        },
`
These registers aren't handled by the driver, so compile fails.
I started off a dts spec for the 'Antmicro' board (Also a litex RiscV eval board), which matches the driver and is the basis for my experiments. 
the regs used there are:
#define UART_RXTX_ADDR          DT_INST_REG_ADDR_BY_NAME(0, rxtx)
#define UART_TXFULL_ADDR        DT_INST_REG_ADDR_BY_NAME(0, txfull)
#define UART_RXEMPTY_ADDR       DT_INST_REG_ADDR_BY_NAME(0, rxempty)
#define UART_EV_STATUS_ADDR     DT_INST_REG_ADDR_BY_NAME(0, ev_status)
#define UART_EV_PENDING_ADDR    DT_INST_REG_ADDR_BY_NAME(0, ev_pending)
#define UART_EV_ENABLE_ADDR     DT_INST_REG_ADDR_BY_NAME(0, ev_enable)
#define UART_TXEMPTY_ADDR       DT_INST_REG_ADDR_BY_NAME(0, txempty)
#define UART_RXFULL_ADDR        DT_INST_REG_ADDR_BY_NAME(0, rxfull)

as you can see 'tuning_word' and 'uart_configured' aren't there.

I haven't checked, how csr.json gets its inputs, so this may be the 'right' thing todo - or write a driver to match the registers - whatever those do... 

In fact some further devices are missing from the generated overlay - in particular the SPI and further on the SDCard.
The scripts even reports some warning, leading to the 'disabling' for specific devices, such as SPI etc.

With my 'derived' dts I'm already on my way to get the SDCard working as well:
The Zephyr 'shell' shows already:
uart:~$ device list
devices:
- serial@f0001000 (READY)
- clock@e0004800 (READY)
- gpio@e0006000 (READY)
- gpio@e0005800 (READY)
- spi@f0002800 (READY)
- sdhc@0 (READY)
  requires: spi@f0002800
- mmc (READY)
  requires: sdhc@0

However, mounting the SDCard (with the other testprog) locks-up somewhere. But at least doesn't report (anymore) that the Card can't be mounted... stay tuned.
(the clock and GPIO devices are still wrong, but I'll look into those later, these are artifacts from the Antmicro board's DTS)

Find attached the intermediate .dts(i) and config files - all likely to change along the fixing of the remaing devices.
Note that I overrule the 'CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=79000000' (orangecart_defconfing). This fixes the time and now is roughly correct (80000000 was still a bit too slow).
Of course it would be the best to generate the DTS and configs from the HW build process, but that's some project for another day... 

So long & happy hacking, martin

[oc-zephyr.zip](https://github.com/zeldin/OrangeCart/files/9895709/oc-zephyr.zip)

**Reply by zeldin** [reply]
Hm?  I don't have `tuning_word` or `configured` in either of `csr.json` or the DTS, but also `uart_liteuart.c` don't use them, so it compiles fine for me.  I got zephyr from git as described on https://docs.zephyrproject.org/latest/develop/getting_started/index.html .

As for SPI, the RVCop64 does not create any register file for SPI flash access, in order to prevent accidental bricking of the bootloader.  I could not find any driver for Litex SDcard in Zephyr, so that might need to be added first.
Disabling devices that you don't have is the correct thing to do.  In my update, I redirect the output from the script to `/dev/null` because it's so noisy about things which are not even wrong.  :-)

If your clock is still wrong, make sure you did a `git submodule update`, since I made a fix in litex to add that define.

**Reply by pottendo** [reply]
Weird, I did the same to clone zephyr.
Did the git submodule update and rebuilt RVCop64:

pottendo@hansi:/work/src/orangeCart/RVCop64/hw(master)$ grep tuning build/*
build/csr.csv:csr_register,uart_tuning_word,0xf0001018,1,rw
build/csr.json:        "uart_tuning_word": {
[...]
build/overlay.dts:        "tuning_word",
build/soc.svd:                            <name>tuning_word</name>

Ad SDCard: well, good to know, that I could brick the device... so I'd better stop!
I've got that far that:
- devices are registered (see above post)
- everything goes well until the first 'send' from SPI at SPI_STATUS_ADDR fails:

zephyr/drivers/spi/spi_litesp.c:83
static void spi_litespi_send(const struct device *dev, uint8_t frame, uint16_t control)
{
[...]
	while (!(litex_read8(SPI_STATUS_ADDR)))
		;
[...]
this while never exits.

Drivers involved:
  zephyr/drivers/spi/spi_litespi.*
  zephyr/drivers/sdhc/sdhc_spi.c

As said, I don't know really how to do this properly. so I'm just wildly copying and trial&error'ing to move forward - so those drivers may be wrong.

So please confirm, that this cannot work, as the registers aren't generated by RVCop64, right?

btw, I mentioned 2 warnings during build - here the details:
pottendo@hansi:/work/src/orangeCart/RVCop64/hw(master)$ grep Warning build.log 
ABC: Warning: The network is combinational.
Warning: Use of IDDRXN and ODDRXN primitives on the same pin is unofficial and unsupported!
Warning: Max frequency for clock '$glbnet$platform_ecp5pll0_clkout0': 65.99 MHz (FAIL at 80.00 MHz)

bye, martin

**Reply by zeldin** [reply]
> Ad SDCard: well, good to know, that I could brick the device... so I'd better stop!

No, SDcard is safe (well, you can destroy the data on your SDcard, but not brick the cartridge).
It's the spiflash pins which are connected to the NOR flash on the PCB that are dangerous in the sense that if you write to the wrong area you can botch the bootloader and need to use JTAG to restore things.

>Warning: Use of IDDRXN and ODDRXN primitives on the same pin is unofficial and unsupported!

This is expected and fine.  It may be "unofficial", but it does work.

> Warning: Max frequency for clock '$glbnet$platform_ecp5pll0_clkout0': 65.99 MHz (FAIL at 80.00 MHz)

This is due to the increased clock frequency, and the reason why 80MHz is not the default.  `nextpnr` can "prove" that the design will work up to 66 MHz, but gives no guarantees for 80 MHz.

**Comment by pottendo** [comment]
Hi fellow hackers,

a brief update for the usage of Zephyr on the RVCop64:
- fixed the slight off-timing
- confirmed pthreads to work (simple tests done)
- C++ stdlib (as provided by Zephyr) works, some weirdness happens, when iostream + floats are used.
- C64 mem-access works
- LED access works, so probably any other GPIO as well
- SPI/SDCard seems working with some tweaks in the Zephyr driver (sdhc, litespi) - see below *)

Left todo is:
- use generated DTS overlay; in my setup the generated overlay isn't working for not-yet investigated reasons
- really check SDCard support and adjust towards a proper driver
- gdb-stub: there's no RiscV gdb-stub in Zephyr yet; not sure if I'll do this.
- find a use-case of Zephyr... ;-)

happy hacking, martin

*) Ad SDCard
I needed to tweak the litespi driver to overrule some control field to get SPI communication running. Additionally some error logs are now disabled. So there's something 'fishy', but besides that, it seems working somehow - need to check more systematic read & write:
```
*** Booting Zephyr OS build zephyr-v3.2.0-843-gd832b04e9627  ***
Starting fs-test...
ioctl...
Block count 15564800Sector size 512
Memory Size(MB) 7600
Disk mounted.

Listing dir /SD: ...
[DIR ] PI1541
[DIR ] 1541
[DIR ] BOOT
[DIR ] DOCUME~1
[DIR ] EXTRA
[DIR ] HARDFP
[DIR ] MODULES
[DIR ] OPT
[FILE] README.MD (size = 1255)
[FILE] CONFIG.TXT (size = 41)
[FILE] KERNEL.IMG (size = 502664)
[FILE] OPTIONS.TXT (size = 5993)
[FILE] BOOTCODE.BIN (size = 52116)
[FILE] FIXUP.DAT (size = 6660)
[FILE] START.ELF (size = 2846852)
[DIR ] TRASH-~1
[FILE] CHARGEN (size = 4096)
[FILE] OPTION~1.TXT (size = 6021)
[FILE] XXXDOS~1 (size = 16384)
[FILE] JIFFY.BIN (size = 16384)
[DIR ] SYSTEM~1
[FILE] REEFER.M64 (size = 1236699136)
[FILE] BIGBUCK.M64 (size = 161855232)
[FILE] ENIGMA.M64 (size = 364892160)
[00:00:02.040,000] <inf> sd: Detected SD card
[00:00:02.100,000] <inf> sd: Maximum SD clock is under 25MHz, using clock of 24000000Hz
[00:00:02.100,000] <inf> main: Block count 15564800
```

**Reply by pottendo** [reply]
hi, 
did a few more investigations on the SDCard support, with the sdhc/sdmmc/litespi driver in Zephyr.
As stated, I needed to tweak the commands and some error handling a bit; I found the hint in the Bios driver. I couldn't find a way to accomplish this without changing some code in zephyr/drivers/spi/litexspi.c - not nice.
With that, reading files seems to work. Writing files fails, as the card seems not to respond properly running either in timeouts or maybe even error responses from the SDCard (one could control this with some 'CONFIG_SD_xxx' macros).

Unfortunately the BIOS Driver doesn't support writing, so I can't figure out what's wrong.
I think I give up here - If anyone has interest in a fully working SDCard under Zephyr, I'd pick up the subject again.
bye, martin
